Kernel/Net/NetworkingManagement.cpp

```diff
@@ -75,16 +85,26 @@ RefPtr<NetworkAdapter> NetworkingManagement::lookup_by_name(const StringView& na
 
 UNMAP_AFTER_INIT RefPtr<NetworkAdapter> NetworkingManagement::determine_network_device(PCI::Address address) const
 {
+#ifdef COMPILE_E1000
     if (auto candidate = E1000NetworkAdapter::try_to_initialize(address); !candidate.is_null())
         return candidate;
+#endif
+#ifdef COMPILE_E1000E
     if (auto candidate = E1000ENetworkAdapter::try_to_initialize(address); !candidate.is_null())
         return candidate;
+#endif
+#ifdef COMPILE_RTL8139
```

Comment:
Seeing the changes to the Networking and Graphics code makes me even less thrilled about the extra #ifdefs added as  a part of this modularization. We discussed a few  options in discord, such as

- keep the ifdefs
- `initalize_if(COMPILE_XXXX)` or similar helpers to hide the incomplete types of non-compiled-in drivers behind an ifdef.
- registering function pointers in the .cpp file for the associated hardware device to perform initialization when requested.
- using global constructors to make FooDeviceX initialization available to FooManagement initialization. 

I'm not sure which one is the best design. These all feel like they could use some design pattern elbow grease. Some factories, or interfaces, or something else to decouple the Managment classes from the concrete implementation classes of each hardware device.

Reply:
So, I tried to find a way to create a `initalize_if(COMPILE_XXXX)` macro, but without much success. I think I saw the nouveau project puts some macros to take another "macro" and use that like the way we want. Still, it becomes not elegant that much to the point I rather put a few ifdefs in the project (and you can see I put a very small amount of them right now) instead of doing that.
So, either we keep the ifdefs and try to find something better in the future or try the last 2 options. The third option is somewhat better imho than the fourth option which I really dislike (I really don't want to mess around with global constructors). Still, I'm not sure I'm 100% positive on how to implement the third option - what will call the registering function I just compiled in the Kernel?

TL;DR - I need a bit more time with this. I might call it a day sometime soon because I don't want to rebase this pull request forever, and it essentially works (minus all the crappiness I did with build files, pardon my lack of knowledge with that sector of the project). However, I don't think ifdefs are great either, so even in the case where we merge this, I'll definitely put some time into this and eventually I'll find a solution to this.

Reply:
Just to put things in good perspective - this PR adds 27 ifdefs to the whole code base. If you look where I added those ifdefs, they are mainly in initialization code (i.e. in Singletons), and they don't have huge blocks within them, like other ifdefs that are already present in the Kernel code... so still, not a great solution, but I'd not treat them as completely and utterly bad pattern if used wisely (so we don't put ifdef on a large block of code... the largest block of code that can be disabled by an ifdef is a block of 9 lines of code right now).

Reply:
I'll give it a bit more time, but I reduced the amount of the ifdefs even more now. It's getting harder for rebasing this pull request as more changes are pushed to master.
Marking this as resolved, although I know this is certainly not the case but we can investigate into this in the near future :)